File: typescript/packages/core/src/core/box/resolve.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { BoxAccessor } from '../../accessor/box.ts'
import type { PathSpec } from '../../types.ts'
import { listFolderItems, type BoxItem, type BoxSearchItem } from './api.ts'

export function pathParts(path: PathSpec): string[] {
  return path.vfsPath.split('/').filter((p) => p !== '')
}

/**
 * Walk folder listings to resolve each component of a path.
 *
 * Box has no path-addressing endpoint, so ids are resolved by listing each
 * level from the mount root. Returns the Box item for every component that
 * resolved, in order; a missing component, or one below a non-folder, ends
 * the chain early.
 */
export async function resolveChain(accessor: BoxAccessor, parts: string[]): Promise<BoxItem[]> {
  const tm = accessor.tokenManager
  let curId = accessor.rootFolderId
  const chain: BoxItem[] = []
  for (const name of parts) {
    if (chain.length > 0 && chain[chain.length - 1]?.type !== 'folder') break
    const children = await listFolderItems(tm, curId)
    const match = children.find((c) => c.name === name)
    if (match === undefined) break
    chain.push(match)
    curId = match.id
  }
  return chain
}

/**
 * Resolve a mount-relative path to its Box item.
 *
 * Returns null if any component is missing, or a non-final component is not
 * a folder.
 */
export async function resolveItem(accessor: BoxAccessor, parts: string[]): Promise<BoxItem | null> {
  const chain = await resolveChain(accessor, parts)
  if (parts.length === 0 || chain.length < parts.length) return null
  return chain[chain.length - 1] ?? null
}

export async function resolveParentId(
  accessor: BoxAccessor,
  parts: string[],
): Promise<string | null> {
  if (parts.length <= 1) return accessor.rootFolderId
  const parent = await resolveItem(accessor, parts.slice(0, -1))
  if (parent?.type !== 'folder') return null
  return parent.id
}

/**
 * Mount-relative path of an item, from its `path_collection`.
 *
 * Box lists an item's ancestors from the account root down to its immediate
 * parent, excluding the item itself; everything up to and including the mount
 * root folder is trimmed. Null when the mount root is not among the
 * ancestors, which is every item outside the mount.
 */
export function mountRelativeKey(
  item: Pick<BoxSearchItem, 'name' | 'path_collection'>,
  rootFolderId: string,
): string | null {
  const entries = item.path_collection?.entries ?? []
  const names: string[] = []
  let collecting = false
  for (const anc of entries) {
    if (collecting) names.push(anc.name)
    if (anc.id === rootFolderId) collecting = true
  }
  if (!collecting) return null
  names.push(item.name)
  return names.filter((n) => n !== '').join('/')
}

File: python/mirage/core/box/resolve.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any

from mirage.accessor.box import BoxAccessor
from mirage.core.box.api import list_folder_items
from mirage.core.box.readdir import ROOT_FOLDER_ID
from mirage.types import PathSpec


def path_parts(path: PathSpec) -> list[str]:
    return [p for p in path.vfs_path.strip("/").split("/") if p]


def root_id(accessor: BoxAccessor) -> str:
    return accessor.config.root_folder_id or ROOT_FOLDER_ID


async def resolve_chain(accessor: BoxAccessor,
                        parts: list[str]) -> list[dict[str, Any]]:
    """Walk folder listings to resolve each component of a path.

    Box has no path-addressing endpoint, so ids are resolved by listing
    each level from the mount root. Returns the Box item for every
    component that resolved, in order; a missing component, or one below
    a non-folder, ends the chain early.

    Args:
        accessor (BoxAccessor): Box accessor.
        parts (list[str]): mount-relative path components.
    """
    tm = accessor.token_manager
    cur_id = root_id(accessor)
    chain: list[dict[str, Any]] = []
    for name in parts:
        if chain and chain[-1].get("type") != "folder":
            break
        children = await list_folder_items(tm, cur_id)
        match = next((c for c in children if c["name"] == name), None)
        if match is None:
            break
        chain.append(match)
        cur_id = match["id"]
    return chain


async def resolve_item(accessor: BoxAccessor,
                       parts: list[str]) -> dict[str, Any] | None:
    """Resolve a mount-relative path to its Box item.

    Returns None if any component is missing, or a non-final component
    is not a folder.

    Args:
        accessor (BoxAccessor): Box accessor.
        parts (list[str]): mount-relative path components.
    """
    chain = await resolve_chain(accessor, parts)
    if not parts or len(chain) < len(parts):
        return None
    return chain[-1]


async def resolve_parent_id(accessor: BoxAccessor,
                            parts: list[str]) -> str | None:
    if len(parts) <= 1:
        return root_id(accessor)
    parent = await resolve_item(accessor, parts[:-1])
    if parent is None or parent.get("type") != "folder":
        return None
    return parent["id"]


def mount_relative_key(item: dict[str, Any],
                       root_folder_id: str) -> str | None:
    """Mount-relative path of an item, from its ``path_collection``.

    Box lists an item's ancestors from the account root down to its
    immediate parent, excluding the item itself; everything up to and
    including the mount root folder is trimmed. None when the mount root
    is not among the ancestors, which is every item outside the mount.

    Args:
        item (dict[str, Any]): A Box item carrying ``path_collection``.
        root_folder_id (str): Box id of the mount root folder.
    """
    entries = (item.get("path_collection") or {}).get("entries") or []
    names: list[str] = []
    collecting = False
    for anc in entries:
        if collecting:
            names.append(anc.get("name", ""))
        if anc.get("id") == root_folder_id:
            collecting = True
    if not collecting:
        return None
    names.append(item.get("name", ""))
    return "/".join(n for n in names if n)
